src/module.c

```diff
@@ -10958,9 +10983,22 @@ void processModuleLoadingProgressEvent(int is_aof) {
     }
 }
 
-/* When a module key is deleted (in dbAsyncDelete/dbSyncDelete/dbOverwrite), it 
+/* When a key is deleted (in dbAsyncDelete/dbSyncDelete/setKey), it
 *  will be called to tell the module which key is about to be released. */
-void moduleNotifyKeyUnlink(robj *key, robj *val, int dbid) {
+void moduleNotifyKeyUnlink(robj *key, robj *val, int dbid, int flags) {
+    server.lazy_expire_disabled++;
+    int subevent = REDISMODULE_SUBEVENT_KEY_DELETED;
+    if (flags & DB_FLAG_KEY_EXPIRED) {
+        subevent = REDISMODULE_SUBEVENT_KEY_EXPIRED;
+    } else if (flags & DB_FLAG_KEY_EVICTED) {
+        subevent = REDISMODULE_SUBEVENT_KEY_EVICTED;
+    } else if (flags & DB_FLAG_KEY_OVERWRITE) {
+        subevent = REDISMODULE_SUBEVENT_KEY_OVERWRITE;
+    }
+    RedisModuleKeyInfoV1 ki = {REDISMODULE_KEYINFO_VERSION, dbid, key};
```

Comment:
Maybe we can give the old value here on the event info? I see we already have it here?

Reply:
you mean so that modules will not have to do RM_OpenKey (improves performance by reducing the need for lookupKey and a memory allocation).

it's interesting, that much of the complexity of what we did (the two phase unlinking), was done so that RM_OpenKey can still succeed from within that callback, so now we make part of that unnecessary. although some parts (like moving the unlinking from the expires dict are anyway necessary).

i guess that if we add a RedisModuleKey member to the struct, we can remove the keyname and dbid ones (we have RM_GetKeyNameFromModuleKey and RM_GetDbIdFromModuleKey)

Reply:
ok, let's give an opened key, users can then get the key name and dbid from that if they want.
i don't think we wanna say that you aren't suppose to do RM_Call, there are many things for which we don't have a direct API for, and the only way to do them is to run a command.
we did do all the work (two phase unlink), in order to make it possible.

we should document that write operations are forbidden from the callback though, but i think we already did.

Reply:
i see we mentioned users are not allowed to modify the key, but maybe instead we should mention that all write operations are forbidden.

p.s. we did add `lazy_expire_disabled` so that read-only access to keys won't introduce deletions.

Reply:
@huangzhw can you make that change too?
i think that's the last thing remaining to merge this (over 1 year old) PR

Reply:
We just use `RM_OpenKey` and `RM_CloseKey`?  So need create a `ctx`?

Reply:
not sure i understand the question.
i think RedisModuleKeyInfoV1 should contain a `RedisModuleKey *key` instead of the current `dbnum` and `key`.
then the user can use RM_GetKeyNameFromModuleKey and RM_GetDbIdFromModuleKey to get them.
you obtain that by calling `moduleInitKey()`

Reply:
We need to create `ctx`, we need a `client` object.

Reply:
yeah, `moduleFireServerEvent` does that for other events as well